providers/ms365/provider/provider.go

```diff
@@ -0,0 +1,242 @@
+
+	"go.mondoo.com/cnquery/llx"
+	"go.mondoo.com/cnquery/providers-sdk/v1/inventory"
+	"go.mondoo.com/cnquery/providers-sdk/v1/plugin"
+	"go.mondoo.com/cnquery/providers-sdk/v1/upstream"
+	"go.mondoo.com/cnquery/providers-sdk/v1/vault"
+	"go.mondoo.com/cnquery/providers/ms365/connection"
+	"go.mondoo.com/cnquery/providers/ms365/resources"
+)
+
+const defaultConnection uint32 = 1
+
+type Service struct {
+	runtimes         map[uint32]*plugin.Runtime
+	lastConnectionID uint32
+}
+
+func Init() *Service {
+	return &Service{
+		runtimes: map[uint32]*plugin.Runtime{},
+	}
+}
+
+func (s *Service) ParseCLI(req *plugin.ParseCLIReq) (*plugin.ParseCLIRes, error) {
+	flags := req.GetFlags()
+
+	tenantId := flags["tenant-id"]
+	clientId := flags["client-id"]
+	clientSecret := flags["client-secret"]
+	certificatePath := flags["certificate-path"]
+	certificateSecret := flags["certificate-secret"]
+
+	opts := map[string]string{}
+	creds := []*vault.Credential{}
+
+	opts["tenant-id"] = string(tenantId.Value)
+	opts["client-id"] = string(clientId.Value)
+
+	if len(clientSecret.Value) > 0 {
+		creds = append(creds, &vault.Credential{
+			Type:   vault.CredentialType_password,
+			Secret: clientSecret.Value,
+		})
+	} else if len(certificatePath.Value) > 0 {
+		creds = append(creds, &vault.Credential{
+			Type:           vault.CredentialType_pkcs12,
+			PrivateKeyPath: string(certificatePath.Value),
+			Password:       string(certificateSecret.Value),
+		})
+	}
+	config := &inventory.Config{
+		Type:        "ms365",
+		Discover:    &inventory.Discovery{Targets: []string{"auto"}},
+		Credentials: creds,
+		Options:     opts,
+	}
+	asset := inventory.Asset{
+		Connections: []*inventory.Config{config},
+	}
+
+	return &plugin.ParseCLIRes{Asset: &asset}, nil
+}
+
+func (s *Service) Connect(req *plugin.ConnectReq, callback plugin.ProviderCallback) (*plugin.ConnectRes, error) {
+	if req == nil || req.Asset == nil {
+		return nil, errors.New("no connection data provided")
+	}
+
+	conn, err := s.connect(req, callback)
+	if err != nil {
+		return nil, err
+	}
+
+	// We only need to run the detection step when we don't have any asset information yet.
+	if req.Asset.Platform == nil {
+		if err := s.detect(req.Asset, conn); err != nil {
+			return nil, err
+		}
+	}
+
+	// discovery assets for further scanning
+	inventory, err := s.discover(conn, conn.Conf)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: discovery of related assets and use them in the inventory below
+	return &plugin.ConnectRes{
+		Id:        uint32(conn.ID()),
+		Name:      conn.Name(),
+		Asset:     req.Asset,
+		Inventory: inventory,
+	}, nil
+}
+
+func (s *Service) connect(req *plugin.ConnectReq, callback plugin.ProviderCallback) (*connection.Ms365Connection, error) {
+	if len(req.Asset.Connections) == 0 {
+		return nil, errors.New("no connection options for asset")
+	}
+
+	asset := req.Asset
+	conf := asset.Connections[0]
+	var conn *connection.Ms365Connection
+	var err error
+
+	switch conf.Type {
```

Comment:
can we remove this?

Reply:
nice catch, also did the same with azure